Source/AwsCommonRuntimeKit/auth/credentials/CRTAWSCredentialsProvider.swift

```diff
@@ -4,51 +4,109 @@
+class DelegateCredentialsProviderCore {
+    let delegateCredentialProvider: CredentialsProvider
+    init(_ credentialsProvider: CredentialsProvider) {
+        self.delegateCredentialProvider = credentialsProvider
+    }
 
-    var rawValue: UnsafeMutablePointer<aws_credentials_provider>
+    func passRetained() -> UnsafeMutableRawPointer {
+        return Unmanaged<DelegateCredentialsProviderCore>.passRetained(self).toOpaque()
+    }
+
+    func release() {
+        Unmanaged.passUnretained(self).release()
+    }
+}
+
+//TODO: Rename file name
+public class AwsCredentialsProvider: CredentialsProvider {
+
+    let allocator: Allocator
+    let rawValue: UnsafeMutablePointer<aws_credentials_provider>
 
     init(credentialsProvider: UnsafeMutablePointer<aws_credentials_provider>,
          allocator: Allocator) {
         self.rawValue = credentialsProvider
         self.allocator = allocator
     }
 
-    public convenience init(fromProvider impl: CRTCredentialsProvider,
-                            allocator: Allocator = defaultAllocator,
-                            shutdownCallback: ShutdownCallback? = nil) throws {
-        let shutdownCallbackCore = ShutdownCallbackCore(shutdownCallback)
-        let credProviderPtr: UnsafeMutablePointer<CRTCredentialsProvider> = fromPointer(ptr: impl)
+    /// Retrieves credentials from a provider by calling its implementation of get credentials and returns them to
+    /// the callback passed in.
+    ///
+    /// - Returns: `Result<CRTCredentials, CRTError>`
+    /// - Throws: CommonRuntimeError.crtError
+    public func getCredentials() async throws -> AwsCredentials {
+        return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<AwsCredentials, Error>) in
+            let continuationCore = ContinuationCore(continuation: continuation)
+            if aws_credentials_provider_get_credentials(rawValue,
+                                                        onGetCredentials,
+                                                        continuationCore.passRetained()) != AWS_OP_SUCCESS {
+                continuationCore.release()
+                continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
+            }
+        }
+    }
+
+    /// Create a credentials provider depends on provided Credential Provider to fetch the credentials.
+    /// - Parameters:
+    ///   - credentialsProvider: The credential provider to use to fetch the credential provider
+    ///   - shutdownCallback:  (Optional) shutdown callback
+    ///   - allocator: (Optional) allocator to override
+    /// - Returns: `CredentialsProvider`
+    /// - Throws: CommonRuntimeError.crtError
+    public static func makeDelegate(credentialsProvider: CredentialsProvider,
+                                    shutdownCallback: ShutdownCallback? = nil,
+                                    allocator: Allocator = defaultAllocator) throws -> AwsCredentialsProvider {
```

Comment:
It's preferable to move this back to an initializer rather than a static function

Reply:
The problem is that initializer can not have names. We can set a name for the first parameter like `fromStatic accessKey: Stirng` but that feels weird because accessKey name should stay the same and not depend upon the constructor. Therefore, we used the factory pattern.